components/brave_sync/brave_profile_sync_service_impl.cc

```diff
@@ -755,6 +766,77 @@ void BraveProfileSyncServiceImpl::SetPermanentNodesOrder(
   brave_sync_prefs_->SetMigratedBookmarksVersion(2);
 }
 
+namespace {
+struct BookmarkByDateAddedComparator {
+  bool operator()(const bookmarks::BookmarkNode* lhs,
+                  const bookmarks::BookmarkNode* rhs) const {
+    DCHECK(lhs);
+    DCHECK(rhs);
+    // TODO(AlexeyBarabash): be ready to empty date added
+    DCHECK(!tools::IsTimeEmpty(lhs->date_added()));
+    DCHECK(!tools::IsTimeEmpty(rhs->date_added()));
+    return lhs->date_added() < rhs->date_added();
+  }
+};
+using SortedNodes =
+    std::set<const bookmarks::BookmarkNode*, BookmarkByDateAddedComparator>;
+using ObjectIdToNodes = std::map<std::string, SortedNodes>;
+
+void FillObjectsMap(const bookmarks::BookmarkNode* parent,
+                    ObjectIdToNodes* object_id_nodes) {
+  for (size_t i = 0; i < parent->children().size(); ++i) {
+    const bookmarks::BookmarkNode* current_child = parent->children()[i].get();
+    std::string object_id;
+    if (current_child->GetMetaInfo("object_id", &object_id) &&
+        !object_id.empty()) {
+      (*object_id_nodes)[object_id].insert(current_child);
+    }
+    if (current_child->is_folder()) {
+      FillObjectsMap(current_child, object_id_nodes);
+    }
+  }
+}
+
+void ClearDuplicatedNodes(ObjectIdToNodes* object_id_nodes,
+                          bookmarks::BookmarkModel* model) {
+  for (ObjectIdToNodes::iterator it_object_id = object_id_nodes->begin();
+       it_object_id != object_id_nodes->end(); ++it_object_id) {
+    const SortedNodes& nodes = it_object_id->second;
+    if (nodes.size() > 1) {
+      // Nodes are sorted from oldest to newest, go to the second by age
+      SortedNodes::iterator it_nodes = nodes.begin();
+      ++it_nodes;
+      for (; it_nodes != nodes.end(); ++it_nodes) {
+        const bookmarks::BookmarkNode* node = *it_nodes;
+        // Copy and delete node, recover its original name
+        const auto original_title = node->GetTitle();
+        const auto* parent = node->parent();
+        size_t original_index = parent->GetIndexOf(node);
+        model->Copy(node, parent, original_index);
+        model->Remove(node);
+        parent->children()[original_index].get()->SetTitle(original_title);
```

Comment:
Why do we have to save title and set it later? Doesn't `BookmarkModel::Copy` also copy the title?

Reply:
I have re-checked, I don't need to restore title, BookmarkModel::Copy also copies the title. Will fix.